python_modules/dagster/dagster/_core/execution/context/invocation.py

```diff
@@ -720,6 +759,105 @@ def set_requires_typed_event_stream(self, *, error_message: Optional[str]) -> No
+
+    def bind(
+        self,
+        op_def: OpDefinition,
+        pending_invocation: Optional[PendingNodeInvocation[OpDefinition]],
+        assets_def: Optional[AssetsDefinition],
+        config_from_args: Optional[Mapping[str, Any]],
+        resources_from_args: Optional[Mapping[str, Any]],
+    ) -> "RunlessAssetExecutionContext":
+        if assets_def is None:
+            raise DagsterInvariantViolationError(
+                "RunlessAssetExecutionContext can only being used to invoke an asset."
+            )
+        if self._op_execution_context._bound_properties is not None:  # noqa: SLF001
+            raise DagsterInvalidInvocationError(
+                f"This context is currently being used to execute {self.op_execution_context.alias}."
+                " The context cannot be used to execute another asset until"
+                f" {self.op_execution_context.alias} has finished executing."
+            )
+
+        self._op_execution_context = self._op_execution_context.bind(
+            op_def=op_def,
+            pending_invocation=pending_invocation,
+            assets_def=assets_def,
+            config_from_args=config_from_args,
+            resources_from_args=resources_from_args,
+        )
+
+        return self
+
+    def unbind(self):
+        self._op_execution_context.unbind()
+
+    @property
+    def bound_properties(self) -> BoundProperties:
+        return self.op_execution_context.bound_properties
+
+    @property
+    def is_bound(self) -> bool:
+        return self.op_execution_context.is_bound
+
+    @property
+    def execution_properties(self) -> RunlessExecutionProperties:
+        return self.op_execution_context.execution_properties
+
+    @property
+    def op_execution_context(self) -> RunlessOpExecutionContext:
+        return self._op_execution_context
+
+    def for_type(self, dagster_type: DagsterType) -> TypeCheckContext:
+        self._check_bound(fn_name="for_type", fn_type="method")
+        resources = cast(NamedTuple, self.resources)
+        return TypeCheckContext(
+            self.run_properties.run_id,
+            self.log,
+            ScopedResourcesBuilder(resources._asdict()),
+            dagster_type,
+        )
+
+    def observe_output(self, output_name: str, mapping_key: Optional[str] = None) -> None:
+        self.op_execution_context.observe_output(output_name=output_name, mapping_key=mapping_key)
+
+    @property
+    def run_properties(self) -> RunProperties:
+        self._check_bound(fn_name="run_properties", fn_type="property")
+        if self._run_props is None:
+            self._run_props = RunlessRunProperties(
+                run_id=self.op_execution_context.run_id,
+                run_config=self.op_execution_context.run_config,
+                # pass None for dagster_run, since RunlessRunProperties raises an exception for this attr
+                dagster_run=None,  # type: ignore
+                retry_number=0,
```

Comment:
We don't handle the retry number in direct invocation right now. If you try to access it via the op context you get an error about accessing the step execution context, since it calls the `retry_number` fn on `OpExecutionContext`. For the asset context, we need to provide a retry number to the `RunProperties` object. We could do this by assuming that the retry number is always 0 (implemented here) or by raising an error on attribute access (like we do with `dagster_run`)

Reply:
also no longer relevant since there is no run properties sub object. still would be good to have a more meaningful error on op context